kube-core/src/version.rs

```diff
@@ -116,19 +177,26 @@ mod tests {
         check_not_parses("v1zeta3");
     }
 
+    #[test]
+    fn test_version_fromstr() {
+        assert_eq!(
+            Version::from_str("infallible").unwrap(),
+            Version::Nonconformant("infallible".to_string())
+        );
+    }
+
     #[test]
     fn test_version_ord() {
-        // NB: semantically this is exact opposite of what makes sense, but we never export Version.
-        // We can reverse the Ord, but would have to remember to .reverse() again after sorts.
-        assert!(Version::Stable(2) < Version::Stable(1));
-        assert!(Version::Stable(2) < Version::Stable(1));
-        assert!(Version::Stable(1) < Version::Beta(1, None));
-        assert!(Version::Stable(1) < Version::Beta(2, None));
-        assert!(Version::Stable(2) < Version::Alpha(1, Some(2)));
-        assert!(Version::Stable(1) < Version::Alpha(2, Some(2)));
-        assert!(Version::Beta(1, None) < Version::Nonconformant("hi".into()));
-
-        // The sorting results is what we export, and this works by default because Ord is reversed:
+        // sorting makes sense from a "greater than" semantic perspective:
+        assert!(Version::Stable(2) > Version::Stable(1));
+        assert!(Version::Stable(1) > Version::Beta(1, None));
+        assert!(Version::Stable(1) > Version::Beta(2, None));
+        assert!(Version::Stable(2) > Version::Alpha(1, Some(2)));
+        assert!(Version::Stable(1) > Version::Alpha(2, Some(2)));
+        assert!(Version::Beta(1, None) > Version::Nonconformant("ver3".into()));
+
+        // sort order by default is ascending
+        // sorting with std::cmp::Reverse thus gives you the "most latest stable" first
```

Comment:
> - reverse Ord impl so we have semantic coherence (stable 2 > stable 1)
> - reverse sort usage internally to compensate (ascending version order default, matches rust conventions)

I think `v2 < v1` is semantic: it means `v2` is sorted before `v1`.

- `v2 < v1`
- `v1 < v1beta1`
- `v1beta1 < v2alpha1`
- `v2alpha1 < a1`
- `a1 < z1`

Is making `v2 > v1` true necessary?

The use of `VersionSortKey` to reverse the ordering of variants' values except nonconformant, and having to use `sort_by_cached_key` with `Reverse` to reverse the order of `VersionSortKey` variants, seems unnecessarily complicated.

Reply:
but that's not what the `>` greater than relation means, semantically; 2 is greater than 1.

we might generally want versions sorted in descending order (and present helpers in discovery that does this), but we should have `v2 > v1`. this is [what the semver crate does](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=894634b1929397776c45ce2e535f40e9)

Reply:
i do think that the way rust makes us do reverse order sorts is pretty complicated though :(

Reply:
For `semver` it's important because it's common to look for a version in range like `< v2.1.0`.

Do we ever need it? Maybe something just enough to allow sorting by latest stable is enough instead of `Version`?

Reply:
discovery module (in `ApiGroup`) needs to present the list in the specified kubernetes version order, but we advocate for using the simpler [`preferred_version_or_latest`](https://docs.rs/kube/latest/kube/discovery/struct.ApiGroup.html#method.preferred_version_or_latest).

It might come in handy in cases where people need to write logic for version migrations across CRDs - and treat everything that's greater than `Version::parse("v2beta2")` as similar enough to do a particular kind of operation. I would certainly be pretty niche, but it is also fully specified how it should behave, so I don't see why we shouldn't implement the order properly.

Reply:
Hmm, I might have been misunderstanding the code. To be clear, I'm not against properly implementing the order at all.

I thought this was mainly needed for sorting the versions (strings) by the latest stable first, and we can delay adding new public `Version` struct until we need them. Like adding a newtype `VersionSortKey(Reverse<Ver>)` that users can use with `.sort_by_cached_key(kube::core::version_sort_key)`.

Reply:
Yeah, that could make sense. I still want to export `Version` though.

But we could make `Version::to_sort_key` public, and maybe also add a `Version::to_sort_key_descending` to make the public sort usage a little less hairy (no need to for users to import `std::cmp::Reverse`)?

Reply:
How common is it actually to need a descending list here? If you want to make an automatic choice then you'd probably just use (`Iterator::max_by_key`)[https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.max_by_key] rather than going via a `Vec`. You can also do a reverse iteration of the `Vec` easily, if that's actually needed (`vec.iter().rev()`).

Reply:
I feel like this is up to the application developers, and not us. They may want a descending/ascending list or they may just want one item (in which case `max_by_key` or `min_by_key` is a great solution). I kind of want a descending list in `kopium` for presentation, and i also want to partition this list at a chosen version to present warnings that there are higher priority versions available that users are not handling.

Reply:
Submitted an alternate proposal FWIW, that tries to clean up `SortVersionKey` and exposes it: https://github.com/kube-rs/kube-rs/pull/767

> I feel like this is up to the application developers, and not us. They may want a descending/ascending list or they may just want one item (in which case max_by_key or min_by_key is a great solution).

Absolutely, my question is just whether this is a case that's common enough that we want to give it special treatment, or whether we should just say "OK, we implement the common traits and it's up to you to decide how to present this".